platform/darwin/test/MGLSymbolStyleLayerTests.mm

```diff
@@ -1096,20 +1098,20 @@ - (void)testProperties {
 
         NSExpression *constantExpression = [NSExpression expressionWithFormat:@"'Text Field'"];
         layer.text = constantExpression;
-        mbgl::style::PropertyValue<std::string> propertyValue = { "Text Field" };
+        mbgl::style::PropertyValue<mbgl::style::expression::Formatted> propertyValue = { "Text Field" };
         XCTAssertEqual(rawLayer->getTextField(), propertyValue,
                        @"Setting text to a constant value expression should update text-field.");
         XCTAssertEqualObjects(layer.text, constantExpression,
                               @"text should round-trip constant value expressions.");
 
-        constantExpression = [NSExpression expressionWithFormat:@"'Text Field'"];
+        constantExpression = [NSExpression expressionWithFormat:@"MGL_FUNCTION('format', 'Text Field', %@)", @{}];
```

Comment:
Is it still possible to set the `textField` property to a string as opposed to a formatted string? If so, we should continue to test that capability alongside the tests for formatted strings. It might require some refactoring of the EJS code that generates this file.

Reply:
I believe that's what the first `layer.text = constantExpression` on line 1100 does? (see ` NSExpression *constantExpression = [NSExpression expressionWithFormat:@"'Text Field'"];`)

Reply:
The workflow I’m concerned about is existing code like this:

```objc
layer.text = [NSExpression expressionForConstantValue:@"Foo"];
// later
if ([layer.text.constantValue isEqualToString:@"Foo"]) { … } // crash
// or
if ([layer.text isEqual:[NSExpression expressionForConstantValue:@"Foo"]]) { … } // false
```

One always takes a risk by directly accessing `NSExpression.constantValue` without first checking the expression type, but if the developer set the `text` property in the first place, they may not expect to have to introduce that check when upgrading to a minor release of the SDK.

In other words, this implicit coercion is unusual in that it’s new. It’s also unusual in that formatted strings are a fairly foreign concept to iOS/macOS development, since we didn’t go the route of using NSAttributedString to represent them. Whereas coercions between strings and numbers are understood as invocations of the `CAST()` operator, there’s no `CAST()` operator support for converting between NSString and whatever class we’d represent a formatted string with. Introducing that `CAST()` support would be an elegant way of at least explaining away the surprise that some developers might experience.

Even so, we know from the v4.0.0 release that any changes to how coercions work can catch developers by surprise. In that case, we’re still fielding questions from developers who don’t expect to have to introduce `CAST()` into their inequality predicates. In this case, on the other hand, developers would be surprised at suddenly having to downcast from formatted strings to strings.

So my suggestion would be to automatically downcast from formatted strings to strings when there’s nothing but a `["formatted", "foo", {}]`. Extra `CAST()` support would be cool too.

Reply:
Right now the getter for _any_ constant expression does the downcast, so the cases you provided all return `true`. This is a little different from doing an automatic downcast based on whether there are options. Even if you add an explicit format expression, you'll still always get a string constant value back. Basically, `format` is almost invisible from the point of view of the iOS SDK with this PR, except that you can set it via `MGL_FUNCTION`.

Test code I used:
```
    MGLSymbolStyleLayer *layer = (MGLSymbolStyleLayer*)[self.mapView.style layerWithIdentifier:@"country-label"];
    layer.text = [NSExpression expressionForConstantValue:@"Foo"];
    // later
    if ([layer.text.constantValue isEqualToString:@"Foo"]) {
        NSLog(@"constantValue comparison worked");
    } // crash
    // or
    if ([layer.text isEqual:[NSExpression expressionForConstantValue:@"Foo"]]) {
        NSLog(@"constant expression comparison worked");
    }
    
    layer.text = [NSExpression expressionWithFormat:@"MGL_FUNCTION('format', 'Bar', %@)", @{}];
    // later
    if ([layer.text.constantValue isEqualToString:@"Bar"]) {
        NSLog(@"constantValue bar comparison worked");
    } // crash
    // or
    if ([layer.text isEqual:[NSExpression expressionForConstantValue:@"Bar"]]) {
        NSLog(@"constant bar expression comparison worked");
    }
```